File: app/javascript/channels/document_channel.js
import consumer from "./consumer"

document.addEventListener("DOMContentLoaded", () => {
  const documentElement = document.getElementById("document_content");
  const UpdateAtElement = document.getElementById("document_last_updated");

  if (documentElement) {
    const id = documentElement.dataset.id;

    const channel = consumer.subscriptions.create(
      { channel: "DocumentChannel", id: id },
      {
        received(data) {
          documentElement.value = data.content;
          UpdateAtElement.textContent = data.updated_at
        },
      }
    );

    documentElement.addEventListener("input", (event) => {
      channel.send({ content: event.target.value });
    });
  }
});

File: app/channels/document_channel.rb
class DocumentChannel < ApplicationCable::Channel
  def subscribed
    stream_from "document_#{params[:id]}"
  end

  def receive(data)
    document = Document.find(params[:id])
    document.update(content: data["content"])
    ActionCable.server.broadcast("document_#{params[:id]}", { content: document.content, updated_at: document.updated_at.to_s })
  end
end
